helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizer.java

```diff
@@ -0,0 +1,347 @@
+              + "] in SwapInstances does not exist in cluster.");
+        }
+      }
+      for (String instanceToRemove : inputFields.instancesToRemove) {
+        if (!liveInstanceSet.contains(instanceToRemove)) {
+          throw new InvalidParameterException("Invalid input: instance [" + instanceToRemove
+              + "] in RemoveInstances does not exist in cluster.");
+        }
+      }
+      if (!inputFields.instancesToRemove.isEmpty()) {
+        clusterState.instances.removeIf(inputFields.instancesToRemove::contains);
+      }
+    }
+
+    // Read instance and cluster config.
+    // It will throw exception if there is no instanceConfig for newly added instance.
+    for (String instance : clusterState.instances) {
+      InstanceConfig config = cfgAccessor.getInstanceConfig(clusterId, instance);
+      clusterState.instanceConfigs.add(config);
+    }
+    clusterState.clusterConfig = cfgAccessor.getClusterConfig(clusterId);
+    return clusterState;
+  }
+
+  private AssignmentResult computeOptimalAssignmentForResources(InputFields inputFields,
+      ClusterState clusterState, String clusterId) throws Exception {
+
+    AssignmentResult result = new AssignmentResult();
+    // Iterate through resources, read resource level info and get potential assignment.
+    HelixDataAccessor dataAccessor = getDataAccssor(clusterId);
+    List<IdealState> wagedResourceIdealState = new ArrayList<>();
+
+    for (String resource : clusterState.resources) {
+      IdealState idealState =
+          dataAccessor.getProperty(dataAccessor.keyBuilder().idealStates(resource));
+      // Compute all Waged resources in a batch later.
+      if (idealState.getRebalancerClassName() != null && idealState.getRebalancerClassName()
+          .equals(WagedRebalancer.class.getName())) {
+        wagedResourceIdealState.add(idealState);
+        continue;
+      }
+      // For non Waged resources, we don't compute resources not in white list.
+      if (!inputFields.resourceFilter.isEmpty() && !inputFields.resourceFilter.contains(resource)) {
+        continue;
+      }
+      // Use getIdealAssignmentForFullAuto for FULL_AUTO resource.
+      if (idealState.getRebalanceMode() == IdealState.RebalanceMode.FULL_AUTO) {
+        String rebalanceStrategy = idealState.getRebalanceStrategy();
+        if (rebalanceStrategy == null || rebalanceStrategy
+            .equalsIgnoreCase(RebalanceStrategy.DEFAULT_REBALANCE_STRATEGY)) {
+          rebalanceStrategy = AutoRebalanceStrategy.class.getName();
+        }
+        Map<String, Map<String, String>> partitionAssignments = new TreeMap<>(
+            getIdealAssignmentForFullAuto(clusterState.clusterConfig, clusterState.instanceConfigs,
+                clusterState.instances, idealState, new ArrayList<>(idealState.getPartitionSet()),
+                rebalanceStrategy));
+        instanceSwapAndFilter(inputFields, partitionAssignments, resource, result);
```

Comment:
Also this function can be called once instead of three times as it's used by each case.

Reply:
`instanceSwapAndFilter` is called for each resources. For non Waged resources, we compute assignment one by one and filter result one by one. Waged resources need to be computed in a batch and filter result one by one. 
I am not sure if these three can be merged.